### PR TITLE
[FSDP] Use unflatten_parameter in _summon_full_parameters

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72467

`_unflatten_params_as_views` does not (and cannot) delete `flat_param`. Using `_unflatten_params_as_views` to implement `_summon_full_parameters` will make the traverse of the full parameter failed -- will get the redundent flat_param.

Differential Revision: [D33989893](https://our.internmc.facebook.com/intern/diff/D33989893/)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 1]
```diff
@@ -1,12 +1,13 @@
 # Owner(s): ["oncall: distributed"]
-import sys
 import math
+import sys
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:3]: nit: Can we have formatting changes in separate diffs in the future, as it makes the diff a lot easier to review :) thanks!

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 2]
```diff
@@ -97,27 +95,25 @@ def test_summon_full_param_shard_value(self):
 
         with model._summon_full_params():
             self.assertEqual(raw_model_size, self.get_model_param_count(model))
-            all_shards = next(model.parameters())
+            parameters = list(model.parameters())
+            all_shards = FlatParameter(parameters, requires_grad=False)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:99]: why do we need to wrap it with `FlatParameter` here?

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 3]
```diff
@@ -271,6 +286,7 @@ def test_params_are_unflatenned(self):
             b = flattened_param.detach()
             self.assertTrue(torch.equal(a, b))
 
+
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:289]: nit: extra line

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 4]
```diff
@@ -197,6 +199,10 @@ def shard_metadata(
             self._sharded_param_offsets[:],
         )
 
+    @property
+    def param_info(self) -> Tuple[ParamInfo]:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flatten_params_wrapper.py:203]: Add a small docstring explaining what this is?

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 5]
```diff
@@ -287,14 +298,17 @@ def _init_flatten_params(
 
         return params, param_infos, shared_param_infos
 
-    def _flatten_params(self) -> None:
+    def _flatten_params(self, external_data: Optional[FlatParameter] = None) -> None:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flatten_params_wrapper.py:301]: Below doc needs to be updated to reflect `external_data` parameter?

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 6]
```diff
@@ -287,14 +298,17 @@ def _init_flatten_params(
...
         """Flatten the managed parameters and replaced the original
         attributes with views to the flat param.
         """
         # register the flatten one
         assert (
-            self.flat_param is not None
-        ), "Can not flatten params when flat_param is None"
+            getattr(self, "flat_param", None) is not None or external_data is not None
+        ), "Can not flatten params when both flat_param and external_data are None."
+        self.flat_param = (
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flatten_params_wrapper.py:309]: Can this be simplified to

```
if external_data:
    self.flat_param = external_data
```

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 7]
```diff
@@ -339,6 +353,26 @@ def _unflatten_params(self) -> None:
 
         del self.flat_param
 
+    @contextlib.contextmanager
+    def unflatten_params(self) -> Generator:
+        """
+        Unflatten params. If the current instance is already unflattened, then
+        it will remain unflattened after the context manager exits.
+        """
+        orig_flattened = self.flat_param is not None
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flatten_params_wrapper.py:362]: Would the code be simpler if we added a check and just returned a noop context manager if orig_flattened = False? Such as

```
orig_flattened = ...
if not orig_flattened:
    return suppress()
# Rest of code
```

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 8]
```diff
@@ -740,16 +743,18 @@ def _summon_full_params(self, recurse: bool = True, writeback: bool = True) -> G
...
 
-            try:
-                yield
-            finally:
-                if writeback:
-                    self._write_back_current_shard()
-                self._free_full_params(currently_local_params)
-                self._use_param_local_shard()
-                self.training_state = TrainingState_.IDLE
+            with contextlib.ExitStack() as stack:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:747]: nit: Add a small comment about what's going on here?

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 9]
```diff
@@ -249,7 +263,9 @@ def test_reshard_outside_forward_backward_iteration(self):
         output = model(torch.zeros(5).cuda(self.rank))
         with model._summon_full_params():
             pass
-        self.assertEqual(outer_full_param_size, outer_param._full_param_padded.storage().size())
+        self.assertEqual(
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:266]: Couldn't see if this was already added, but can we add a test that the parameter names are fixed / back to the original module's with this unflattening fix?

**Review by rohan-varma — COMMENTED** [review]
Looks great overall! Just some small comments/questions before stamping

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 10]
```diff
@@ -243,6 +249,11 @@ def __init__(self, module: nn.Module, param_list: List[nn.Parameter]):
         self.flat_param = FlatParameter(params, params[0].requires_grad)
         self.flat_param._param_infos = param_infos
         self.flat_param._shared_param_infos = shared_param_infos
+        # This attribute is used to remember the flat_param inside the
+        # unflatten_params() context. Using an attribute instead of a
+        # local variable allows us to access the flat_param metadata
+        # even if flat_param is temporarily deleted.
+        self._orig_flat_param = [None]
```
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/distributed/fsdp/flatten_params_wrapper.py:256]: curious why local variable '_orig_flat_param' can not access the flat_param object after self.flat_param is deleted? iiuc, local variable '_orig_flat_param' will still point to the object, 'del self.flat_param' is just reduce the reference counts of the object.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 11]
```diff
@@ -740,16 +743,18 @@ def _summon_full_params(self, recurse: bool = True, writeback: bool = True) -> G
 
             currently_local_params = self._collect_local_params()
             self._rebuild_full_params()
-            self._fsdp_wrapped_module._unflatten_params_if_needed()
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:743]: It's still used in forward.

File: test/distributed/fsdp/test_fsdp_summon_full_params.py (hunk 9 above)
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_summon_full_params.py:267]: I added a test to check the number of `fsdp_model.parameters()` and the value of the parameters. Testing keys is more tricky as this relies on `state_dict()`  or `named_parameter()` which both do not call `_summon_full_parameters()` in the current code base.

Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_summon_full_params.py:267]: I see, I filed an issue to track this to make sure that we fix it: https://github.com/pytorch/pytorch/issues/72716

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 12]
```diff
@@ -271,6 +287,18 @@ def test_params_are_unflatenned(self):
             b = flattened_param.detach()
             self.assertTrue(torch.equal(a, b))
 
+    def test_params_count_and_value(self):
+        model = nn.Sequential(
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:291]: you can also import and test `NestedWrappedModel` to test a nested model.

**Review by rohan-varma — APPROVED** [review]
LGTM

Please ensure @zhaojuanmao 's comments are resolved as well :)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:743]: I see, keep it then

File: torch/distributed/fsdp/flatten_params_wrapper.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/distributed/fsdp/flatten_params_wrapper.py]: that makes sense!

**Comment by zhaojuanmao** [comment]
lgtm as well, thanks for fixing it!

**Pull request closed without merging**